### ISSUE TITLE
starts, then fail

### ISSUE THREAD

**Issue by phiweger** [kblin/ncbi-genome-download#29]
Hi,

I try to download all viral reference genomes in refseq like so:

```
ncbi-genome-download --format fasta --section refseq --verbose --parallel 10 --human-readable viral
```

The program starts downloading things but then, at the same ID (GCF...) breaks with the following error message:

```
INFO: Downloading record 'GCF_000868825.1'
INFO: Downloading record 'GCF_000844105.1'
Traceback (most recent call last):
  File "/Users/pi/.virtualenvs/lab3/bin/ncbi-genome-download", line 11, in <module>
    sys.exit(main())
  File "/Users/pi/.virtualenvs/lab3/lib/python3.5/site-packages/ncbi_genome_download/__main__.py", line 73, in main
    ret = ncbi_genome_download.download(args)
  File "/Users/pi/.virtualenvs/lab3/lib/python3.5/site-packages/ncbi_genome_download/core.py", line 60, in download
    args.taxid, args.human_readable, args.parallel)
  File "/Users/pi/.virtualenvs/lab3/lib/python3.5/site-packages/ncbi_genome_download/core.py", line 75, in _download
    for entry in entries:
  File "/Users/pi/.virtualenvs/lab3/lib/python3.5/site-packages/ncbi_genome_download/summary.py", line 28, in __next__
    entry[self._fields[i]] = val
IndexError: list index out of range
```

I tried a minimal version as well, same result:

```
ncbi-genome-download --verbose --parallel 10 viral
```

Am I using it wrongly or is this an uncaught error, like some empty entry in the database or similar.

Thanks a lot!

**Comment by kblin** [comment]
I can see the error here as well, I'll investigate.
Thanks for reporting this.

**Comment by kblin** [comment]
On a first look, it seems like there's an extra tab in the `assembly_summary.txt` file for record `GCF_000867225.19`, causing it to not have the right amount of fields.

Let me see if I can make the code more robust to just print an error and skip such lines, instead of crashing.

**Comment by viehwegerlib** [comment]
That would be awesome, thx!

**Comment by kblin** [comment]
Wow, this seems to affect a lot of lines in the viral assembly_summary.txt file

**Comment by kblin** [comment]
Currently checking if my fix successfully downloads the whole set of viral entries.

**Comment by kblin** [comment]
Of the 7090 entries in the viral `assembly_summary.txt` file, 694 have this error.

**Comment by viehwegerlib** [comment]
This seems plausible.

**Comment by kblin** [comment]
Should be fixed in the current git master branch.

**Comment by viehwegerlib** [comment]
Installed via

```
pip install git+https://github.com/kblin/ncbi-genome-download.git@master
```

and works like a charm.

```
...
INFO: Downloading record 'GCF_000847025.1'
INFO: Downloading record 'GCF_000837805.1'
INFO: Downloading record 'GCF_000847085.1'
ERROR: Invalid line length in summary file line 129. Expected 21, got 22. Skipping entry.
ERROR: Invalid line length in summary file line 130. Expected 21, got 22. Skipping entry.
ERROR: Invalid line length in summary file line 131. Expected 21, got 22. Skipping entry.
ERROR: Invalid line length in summary file line 132. Expected 21, got 22. Skipping entry.
ERROR: Invalid line length in summary file line 133. Expected 21, got 22. Skipping entry.
ERROR: Invalid line length in summary file line 134. Expected 21, got 22. Skipping entry.
INFO: Downloading record 'GCF_000852145.1'
INFO: Downloading record 'GCF_000847585.1'
...
```

I close this issue and would award you an open source quick response knightship if there were one!

**Comment by viehwegerlib** [comment]
The code now skips "22-tab" entries. But are they valid entries, i.e. as concerns sequence data? Wouldn't it then be better to ignore the column instead of the entire entry?

**Comment by kblin** [comment]
Reopening issues where fix hasn't been shipped yet.

**Comment by kblin** [comment]
I've only looked at the first invalid entry, and that one has the "download URL" shifted into the wrong field. I'm not sure if they all have their invalid tab in the same place, so figuring out what needs to go where might be a bit tricky. But the download URL not being in the right place kind of kills the entry from a downloading tool perspective.

I have not followed up checking if the sequencing data looks sane, I assume it is, I see no reason why it wouldn't.

**Comment by kblin** [comment]
Ok, on a second look, it _seems_ to be consistent. Let me run some more tests. :)

**Comment by kblin** [comment]
Ok, I think I can fix pretty much all the affected entries. My test download is still running, but you should be able to fetch the skipped entries by just rerunning `ncbi-genome-download` in the latest git master version.

**Comment by kblin** [comment]
Yes, it's down to one skipped entry now, in line 7060. That line just has one field too many, again breaking the ftp_path field. I'm not really sure what the "proper" fix would be for this one. NCBI just needs to fix their file. :)

**Comment by kblin** [comment]
Shipped in 0.2.3 release.

**Comment by viehwegerlib** [comment]
Pure awesomeness!

**Comment by andrewsanchez** [comment]
I have also seen this error recently on version 0.2.9:

```
ncbi-genome-download -v -H -F fasta -p 8 -s genbank bacteria

INFO: Checking record 'GCA_003482455.1'
INFO: Checking record 'GCA_003504155.1'
INFO: Checking record 'GCA_003543135.1'
INFO: Checking record 'GCA_002401225.1'
Traceback (most recent call last):
  File "/home/asanchez/miniconda3/envs/genome-download/bin/ncbi-genome-download", line 11, in <module>
    sys.exit(main())
  File "/home/asanchez/miniconda3/envs/genome-download/lib/python3.6/site-packages/ncbi_genome_download/__main__.py", line 24,
in main
    ret = args_download(args)
  File "/home/asanchez/miniconda3/envs/genome-download/lib/python3.6/site-packages/ncbi_genome_download/core.py", line 151, in
args_download
    return config_download(config)
  File "/home/asanchez/miniconda3/envs/genome-download/lib/python3.6/site-packages/ncbi_genome_download/core.py", line 184, in
config_download
    download_jobs.extend(create_downloadjob(entry, group, config))
  File "/home/asanchez/miniconda3/envs/genome-download/lib/python3.6/site-packages/ncbi_genome_download/core.py", line 343, in create_downloadjob
    symlink_path = create_readable_dir(entry, config.section, domain, config.output)
  File "/home/asanchez/miniconda3/envs/genome-download/lib/python3.6/site-packages/ncbi_genome_download/core.py", line 387, in create_readable_dir
    get_species_label(entry),
  File "/home/asanchez/miniconda3/envs/genome-download/lib/python3.6/site-packages/ncbi_genome_download/core.py", line 560, in get_species_label
    return entry['organism_name'].split(' ')[1]
IndexError: list index out of range

[asanchez@bio653 ~/databases ]$ ncbi-genome-download --version
0.2.9
```

Any ideas?

**Comment by jrjhealey** [comment]
Does it always fail on the same accession?

**Comment by andrewsanchez** [comment]
That appears to be the case.   I'm running it again with:

```python
    try:
        return entry['organism_name'].split(' ')[1]
    except IndexError:
        import pdb; pdb.set_trace()
```

**Comment by andrewsanchez** [comment]
https://www.ncbi.nlm.nih.gov/assembly/GCA_002401225.1/

```
$ sed -n 2p genbank_bacteria_assembly_summary.txt >> GCA_002401225.1.csv
$ grep GCA_002401225.1 genbank_bacteria_assembly_summary.txt >> GCA_002401225.1.csv
$ cat GCA_002401225.1.csv
# assembly_accession    bioproject      biosample       wgs_master      refseq_category taxid   species_taxid   organism_name infraspecific_name       isolate version_status  assembly_level  release_type    genome_rep      seq_rel_date    asm_name      submitter        gbrs_paired_asm paired_asm_comp ftp_path        excluded_from_refseq    relation_to_type_material
GCA_002401225.1 PRJNA391950     SAMN07568955    NVSM00000000.1  na      1869227 1869227 bacterium               NORP141 latestContig   Major   Full    2017/09/28      ASM240122v1     University of Southern California       na      na      ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002/401/225/GCA_002401225.1_ASM240122v1     derived from metagenome
```

Running without `-H` flag results in a succesful download:

```
$ ncbi-genome-download -s genbank -A GCA_002401225.1 bacteria
$ lls genbank/bacteria/GCA_002401225.1/
total 1876
-rw-rw-r-- 1 asanchez 1911419 Mar 22 12:30 GCA_002401225.1_ASM240122v1_genomic.gbff.gz
-rw-rw-r-- 1 asanchez    1215 Mar 22 12:30 MD5SUMS
```

But this didn't:
```
$ ncbi-genome-download -v -H -F fasta -s genbank -A GCA_002401225.1 bacteria
```

Maybe do this?
```python
    try:
        return entry['organism_name'].split(' ')[1]
    except IndexError:
        return entry['organism_name']
```

If I remember correctly, `organism_name` is inconsistent and unreliable anways.  For this genome,  `organism_name` is "baceterium."   I've gotten much better results by first downloading  `ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz` and getting the "scientific names" from that file by looking up the `species_taxid`.

**Comment by kblin** [comment]
Thanks for the great analysis (and the patch in #87). I'll reopen the issue until the fix is in a release.

**Comment by andrewsanchez** [comment]
Also, perhaps an even better way to create the mapping between IDs and the standardized "scientific name," (as opposed to using the unstandardized "organism_name" from `assembly_summary.txt` is the following script that uses NCBI's Entrez tools:

```
# Get values from species_taxid (column 7)
# Skip the first two lines which have a comment and column names
cut -f 7 assembly_summary.txt | tail -n +3 |
sort | uniq > species_taxids.txt

epost -db taxonomy -input species_taxids.txt | \
    efetch -format docsum | \
        xtract -pattern DocumentSummary \
         -group DocumentSummary -element Id -element ScientificName \
         >> scientific_names.tsv
```

**Comment by kblin** [comment]
I would like to not hit the network that often, and would also prefer to not depend on the entrez tools during run-time.

**Comment by jrjhealey** [comment]
@andrewsanchez you should be able to essentially get that information from the slightly more verbose output of `gimme_taxa` which gets the tax dump in one go rather than hammering Entrez

**Comment by jrjhealey** [comment]
@andrewsanchez  if you just want to go from TaxID's to the 'full lineage', IIRC, the `gimme_taxa.py` script will do this with the verbose output.

If it doesn't, it would be fairly trivial to modify it so that it does, which is essentially what I did in: https://github.com/jrjhealey/PYlogeny

**Comment by kblin** [comment]
The IndexError is fixed in the 0.2.10 release.
